### PR TITLE
[3.5] Astar bitmask layers and `set_as_bulk_array()` for Astar

### PR DESCRIPTION
**Pull request by KnightNine**
This is made for the 3.5 branch because I'm only using stable branches of Godot. Maybe someone else who wants this can edit it for the master branch, do with this what you will.

- Adds parallel Astar bitmask layer functionality to specify what layers of the Astar map you want to use when pathing.
- adds a `set_as_bulk_array()` function which may or may not speed up loading large astar maps. (haven't tested the performance increase yet), this merge request is mostly for the layer functionality.

Doesn't work with Astar 2D as of yet.

here's some basic code I've tested in gdscript so far to demonstrate the main functionality:
```gdscript
var pool_points = PoolRealArray()
var max_connections = 1
var pool_connections = PoolIntArray()
func append_to_pool_points(arr):
	for f in arr:
		pool_points.append(f)
func append_to_pool_connections(arr):
	for i in arr:
		pool_connections.append(i)


func _ready():
	
	
	
	var support_layers0 = 1<<1
	var support_layers1 = (1<<2) + (1<<1)
	OS.alert(str(support_layers0)+","+str(support_layers1))
	
	var line_points = 4
	
	var astar = AStar.new()
	var id = 0
	
	#start and end should support both layer indexes (support_layers1)
	
	#add a point at the top and bottom for end and start points
	#bottom
	append_to_pool_points([id,1,-1,0,1,support_layers1])
	id+=1
	
	#top
	append_to_pool_points([id,1,line_points+1,0,1,support_layers1])
	id+=1
	
	#create two lines between these top and bottom points
	
	
	var connect_to_last = false
	for i in range(line_points):
		append_to_pool_points([id,1,i,0,1,support_layers0])
		if connect_to_last:
			print(str(id)+ "<->"+str(id-1))
			
			append_to_pool_connections([id,id-1])
		else:
			print(str(id)+ "<->"+str(0))
			#connect to bottom
			append_to_pool_connections([id,0])
		
		connect_to_last = true
		id+=1
	
	#connect to top
	append_to_pool_connections([id-1,1])
	
	var arc_index = 0
	
	#adjacent line that is bent in the x axis to be a longer path
	connect_to_last = false
	for i in range(line_points):
		if i < line_points/2.0:
			arc_index += 1
		else:
			arc_index -= 1
		
		append_to_pool_points([id,-1 * (1+max(arc_index,0)),i,0,1,support_layers1])
		
		if connect_to_last:
			print(str(id)+ "<->"+str(id-1))
			append_to_pool_connections([id,id-1])
		else:
			print(str(id)+ "<->"+str(0))
			#connect to bottom
			append_to_pool_connections([id,0])
		
		connect_to_last = true
		
		id+=1
		
	
	#connect to top
	append_to_pool_connections([id-1,1])
	
	#now test loading this stored astar map data as a bulk array
	astar.set_as_bulk_array(pool_points,max_connections,pool_connections)
	
	#now try pathing 
	#this path should be forced to take the longer path which supports layer 2
	var path = astar.get_id_path(0,1,(1<<2))
	OS.alert(str(path))
	#while this path is free to use the shorter path since it supports layer 1
	var path1 = astar.get_id_path(0,1,support_layers0)
	OS.alert(str(path1))
	#this path can use all layers
	var path2 = astar.get_id_path(0,1,0)
	OS.alert(str(path2))
	
	
```

issues:
- wasn't 100% sure that the max bitmask layer value is equal to `1 << 31 -1` or if it is less than or more than this value. 

This is related to my proposals [here](https://github.com/godotengine/godot-proposals/issues/5188) and [here](https://github.com/godotengine/godot-proposals/issues/4658)

### CODE REVIEW

**Comment by Chaosus** [comment]
You need to squash the commits as required by our pipeline.

The general rule is the changes should be made for the master branch first. Not sure if this change is really useful or easy to use for the other users (added to the PR review meeting).

**Comment by Zireael07** [comment]
+1 to bulk arrays/pooling, but I'm kinda confused on what the bitmask layers achieve. This needs docs written at the same time as the feature is implemented...

**Comment by KnightNine** [comment]
> You need to squash the commits as required by our pipeline.

not sure how to do that but I'll figure it out sometime.

> The general rule is the changes should be made for the master branch first.

yes I know, but I'm ~~lazy~~ busy.

> Not sure if this change is really useful or easy to use for the other users.

Well it requires the user to know how bitmasks work, you could instead have the user input an array of the relevant layers desired and convert that into a bitmask value for the `add_point` and `create_point_path` functions. Though the user would still need to store the bulk arrays using the bitmask values and I think it's better to have that consistency for what the "layers" parameter represents.

Is it useful? I mean it's probably a pretty niche purpose, but If anyone has two or more entities which use the same astar map but with slight variations for any reason then this will save all the time of implementing, storing, and then loading multiple duplicate astar maps. 

Bruh it was taking me literally 70 seconds to load 8 variations of the same Astar map when I started my project.




> but I'm kinda confused on what the bitmask layers achieve. This needs docs written at the same time as the feature is implemented...

Sorta explained them in the `Astar.xml` file. but maybe more detail is needed.

File: core/math/a_star.cpp [hunk 1]
```diff
@@ -69,6 +71,120 @@ void AStar::add_point(int p_id, const Vector3 &p_pos, real_t p_weight_scale) {
...
 }
 
+void AStar::append_as_bulk_array(const PoolVector<real_t> &pool_points, int max_connections, const PoolVector<int> &pool_connections)
+{
+	
+
+	
+
+
+
```
Comment (Chaosus) [inline core/math/a_star.cpp:81]: You need to use `clang-format` tool to avoid such incorrect spaces.

**Comment by Chaosus** [comment]
> not sure how to do that but I'll figure it out sometime.

I think you should read the https://docs.godotengine.org/en/latest/community/contributing/pr_workflow.html first. Be careful to do the change like this using the console - better use the Git applications or tools to do that squashing.

> Is it useful? I mean it's probably a pretty niche purpose, but If anyone has two or more entities which use the same astar map but with slight variations for any reason then this will save all the time of implementing, storing, and then loading multiple duplicate astar maps.

Well, it may be really useful, but may be wrongly implemented - that's why it should be properly reviewed.

**Comment by kleonc** [comment]
> This is made for the 3.5 branch because I'm only using stable branches of Godot. Maybe someone else who wants this can edit it for the master branch, do with this what you will.

Besides the lacking `master` version this PR should be against `3.x`, not `3.5`. I think it could still be cherry-picked for `3.5.x` as a _"backwards-compatible usability enhancement"_ which are allowed according to [the release policy](https://docs.godotengine.org/en/3.5/about/release_policy.html#release-support-timeline).

---

After brief look: layers is a bitmask so `0` should mean no layer, not all layers. All layers are represented by all bits set (`-1` for signed integer). And the default values should be consistent with the rest of the engine and seems like other things default to layer 1.

I also dislike the `parallel_support_layers` naming. Long and I have no idea what it means. I think something like `point_layers` would be better.

**Comment by smix8** [comment]
This should be split into 2 pr's as there are also two different proposals, one for AStar class navigation_layers and one for AStar bulk_functions. The later I don't support, at least not in this current implementation.

As mentioned in the navigation_layers proposal the AStar classes should get "navigation_layers". That is the used term across all navigation nodes and objects for the same bitmask functionality. We are not using "parallel_support_layers" as that term means nothing to users.

The navigation_layers need to be 1 as default like all other layers as 0 is a disabled bitmask which would disable the pathfinding.

Every navigation change needs to be done for master first and after some testing can be backported for Godot 3.5.x / 3.6.

All the other already mentioned stuff like proper doc and clang-format needs to be done as well.

**Comment by KnightNine** [comment]
"point layers" is a bit redundant since the variable is already contained within the point class. Maybe "navigation layers" is also redundant since AStar is used for... navigation. So maybe you should just call them "layers".

> The later I don't support, at least not in this current implementation.
understandable but do you have an idea of how this could be implemented any other way? Is there a way to mass create a bunch of objects from the Point class all at the same time, and then load all of those points' connections all at the same time without looping through them?
maybe how the points are stored needs to be entirely reworked.

> Look up hierarchical pathfinding and be clever about your problem instead of doubling down with a "bulk function". That long loading times should give you a hint that you have way to many graph points for runtime pathfinding performance.

yes and no, HPA* is probably not a good idea for 3D astar chunks since i'd have a lot of connections between each chunk and i'd need to recalculate the "entrance to exit" distances whenever some entity (which occupies/disables points) enters the chunk.
I'll need to think more about how I can potentially make use of HPA*.

I was thinking about the inverse of this by [removing points from chunks which are empty and allowing direct paths between the edges of these empty spaces while they are empty](https://godotforums.org/d/30875-trying-to-optimize-astar-usage) to minimize the number of graph points in the air (for flying entities) but godot's Astar implementation has no idea what a "direct path" is since it isn't tied to any sort of grid.

**Comment by kleonc** [comment]
> "point layers" is a bit redundant since the variable is already contained within the point class. Maybe "navigation layers" is also redundant since AStar is used for... navigation. So maybe you should just call them "layers".

I meant how it should be referred to in the exposed API. From the user perspective there's only `AStar2D/3D` class (internal `Point` class is not exposed) and these layers are only a parameter in some methods. And in the current implementation these layers are specifically point (vertex) layers. They  don't apply to connections (edges). Which leads to a question: why actually only vertices have layers but not edges? Currently for example you can't have two _walkable_ vertices with a non-_walkable_ connection between them.

**Comment by smix8** [comment]
@kleonc There is another proposal / issue that I remember but I cannot find about that. AStar classes currently have the nuisance for many game types that there is no distinction between edge movement cost and the vertex cost or the connections which leads to such practical problems and silly pathfinding situations.

**Comment by Calinou** [comment]
Thanks for opening a pull request :slightly_smiling_face: 

Pull requests targeting the stable branch should be opened against the `3.x` branch instead of `3.5`. This is because feature development for the stable branch occurs in the `3.x` branch.

**Comment by akien-mga** [comment]
Closing as per above comment, this would need to target the `3.x` branch so it needs to be redone.

**Pull request closed without merging**